File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
n = int(input())
s = [int(input()) for _ in range(n)]
d = [[0,0] for _ in range(n)]

for i in range(n):
    if i==0:
        d[i][0]=s[i]
        continue
    if i==1:
        d[i][0]=s[1]
        d[i][1]=s[0]+s[1] 
        continue
    d[i][0] = max(d[i-2][0],d[i-2][1]) + s[i]
    d[i][1] = d[i-1][0]+s[i]

print(max(d[n-1]))

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N+1];

        for(int i=1; i<=N; i++){
            arr[i] = Integer.parseInt(br.readLine());
        }

        int[][] dp = new int[N+1][3];
        dp[1][1] = arr[1];
        for(int i=2; i<=N; i++){
            dp[i][0] = Math.max(dp[i-1][1], dp[i-1][2]);
            dp[i][1] = arr[i] + dp[i-1][0];
            dp[i][2] = arr[i] + dp[i-1][1];
        }

        System.out.println(Math.max(dp[N][1], dp[N][2]));
    }
}